Submitted By: Robert Connolly <cendres at videotron dot ca> (ashes)
Date: 2003-12-01
Initial Package Version: 2.6.0
Origin: http://www.northernsecurity.net/adamantix/diffs/2.6.0-propolice.patch
Description: Linux kernel patch for ProPolice Stack Gaurd
http://www.linuxfromscratch.org/hints/downloads/files/winter.txt
http://www.topside.org/~ashes/files/protector/winter.txt

diff -uNr linux-2.6.0-test9-bk1/arch/i386/kernel/i386_ksyms.c linux-2.6.0-test9-bk1-ssp/arch/i386/kernel/i386_ksyms.c
--- linux-2.6.0-test9-bk1/arch/i386/kernel/i386_ksyms.c	2003-10-25 20:44:54.000000000 +0200
+++ linux-2.6.0-test9-bk1-ssp/arch/i386/kernel/i386_ksyms.c	2003-10-28 14:16:45.000000000 +0100
@@ -97,6 +97,11 @@
 EXPORT_SYMBOL_NOVERS(__down_failed_interruptible);
 EXPORT_SYMBOL_NOVERS(__down_failed_trylock);
 EXPORT_SYMBOL_NOVERS(__up_wakeup);
+
+/* SSP */
+EXPORT_SYMBOL_NOVERS(__guard);
+EXPORT_SYMBOL_NOVERS(__stack_smash_handler);
+
 /* Networking helper routines. */
 EXPORT_SYMBOL(csum_partial_copy_generic);
 /* Delay loops */
diff -uNr linux-2.6.0-test9-bk1/arch/um/kernel/ksyms.c linux-2.6.0-test9-bk1-ssp/arch/um/kernel/ksyms.c
--- linux-2.6.0-test9-bk1/arch/um/kernel/ksyms.c	2003-10-25 20:42:51.000000000 +0200
+++ linux-2.6.0-test9-bk1-ssp/arch/um/kernel/ksyms.c	2003-10-28 14:16:45.000000000 +0100
@@ -90,3 +90,5 @@
 EXPORT_SYMBOL(kmap_atomic_to_page);
 #endif
 
+EXPORT_SYMBOL_NOVERS(__guard);
+EXPORT_SYMBOL_NOVERS(__stack_smash_handler);
diff -uNr linux-2.6.0-test9-bk1/include/linux/kernel.h linux-2.6.0-test9-bk1-ssp/include/linux/kernel.h
--- linux-2.6.0-test9-bk1/include/linux/kernel.h	2003-10-25 20:42:42.000000000 +0200
+++ linux-2.6.0-test9-bk1-ssp/include/linux/kernel.h	2003-10-28 14:16:46.000000000 +0100
@@ -110,6 +110,8 @@
 #define TAINT_FORCED_RMMOD		(1<<3)
 
 extern void dump_stack(void);
+extern int __guard;
+extern void __stack_smash_handler(int, char []);
 
 #ifdef DEBUG
 #define pr_debug(fmt,arg...) \
diff -uNr linux-2.6.0-test9-bk1/lib/Makefile linux-2.6.0-test9-bk1-ssp/lib/Makefile
--- linux-2.6.0-test9-bk1/lib/Makefile	2003-10-25 20:42:51.000000000 +0200
+++ linux-2.6.0-test9-bk1-ssp/lib/Makefile	2003-10-28 14:16:46.000000000 +0100
@@ -5,7 +5,7 @@
 
 lib-y := errno.o ctype.o string.o vsprintf.o cmdline.o \
 	 bust_spinlocks.o rbtree.o radix-tree.o dump_stack.o \
-	 kobject.o idr.o div64.o parser.o
+	 kobject.o idr.o div64.o parser.o propolice.o
 
 lib-$(CONFIG_RWSEM_GENERIC_SPINLOCK) += rwsem-spinlock.o
 lib-$(CONFIG_RWSEM_XCHGADD_ALGORITHM) += rwsem.o
diff -uNr linux-2.6.0-test9-bk1/lib/propolice.c linux-2.6.0-test9-bk1-ssp/lib/propolice.c
--- linux-2.6.0-test9-bk1/lib/propolice.c	1970-01-01 01:00:00.000000000 +0100
+++ linux-2.6.0-test9-bk1-ssp/lib/propolice.c	2003-10-28 14:16:46.000000000 +0100
@@ -0,0 +1,13 @@
+ /*
+  *  linux/lib/errno.c
+  *
+  *
+  */
+ 
+ int __guard = '\0\0\n\777';
+ 
+ void __stack_smash_handler (int damaged, char func[])
+ 	 {
+	   static char *message = "propolice detects %x at function %s.\n" ;
+	   panic (message, damaged, func);
+}
diff -uNr linux-2.6.0-test9-bk1/Makefile linux-2.6.0-test9-bk1-ssp/Makefile
--- linux-2.6.0-test9-bk1/Makefile	2003-10-28 14:21:32.000000000 +0100
+++ linux-2.6.0-test9-bk1-ssp/Makefile	2003-10-28 14:17:04.000000000 +0100
@@ -276,7 +276,7 @@
 		   $(if $(KBUILD_SRC),-Iinclude2 -I$(srctree)/include)
 
 CFLAGS 		:= -Wall -Wstrict-prototypes -Wno-trigraphs -O2 \
-	  	   -fno-strict-aliasing -fno-common
+	  	   -fno-strict-aliasing -fno-common -fstack-protector
 AFLAGS		:= -D__ASSEMBLY__
 
 export	VERSION PATCHLEVEL SUBLEVEL EXTRAVERSION KERNELRELEASE ARCH \
